src/storage/src/source/source_reader_pipeline.rs

```diff
@@ -838,7 +859,70 @@ where
         scope.clone(),
         move |mut _capabilities, _frontiers, scheduler| async move {
             let mut buffer = Vec::new();
+
+            let mut persist_clients = persist_clients.lock().await;
+            let persist_client = persist_clients
+                .open(storage_metadata.persist_location.clone())
+                .await
+                .expect("error creating persist client for Healthchecker");
+            drop(persist_clients);
+
+            let client_ref = &persist_client;
+
+            let write_to_persist = |row: Row, timestamp: Timestamp| async move {
+                let status_shard = match storage_metadata.status_shard {
+                    Some(s) => s,
+                    None => return,
+                };
+                let handle = client_ref
+                    .open_writer(status_shard)
+                    .await;
+
+                let mut handle = match handle {
+                    Ok(stuff) => stuff,
+                    Err(e) => {
+                        panic!("Invalid usage of the persist client for source {source_id} status history shard: {e:?}");
+                    }
+                };
+
+                let mut recent_upper = handle.upper().clone();
+                let mut append_ts = timestamp;
+                'retry_loop: loop {
+                    // We don't actually care so much about the timestamp we append at; it's best-effort.
+                    // Ensure that the append timestamp is past the current upper, and the new upper
+                    // is past that timestamp. (Unless we've advanced to the empty antichain, but in
+                    // that case we're already in trouble.)
+                    for t in recent_upper.elements() {
+                        append_ts.join_assign(t);
+                    }
+                    let mut new_upper = Antichain::from_elem(append_ts.step_forward());
+                    new_upper.join_assign(&recent_upper);
+
+                    let updates = vec![
+                        ((SourceData(Ok(row.clone())), ()), append_ts, 1i64)
+                    ];
+                    let cas_result = handle.compare_and_append(updates, recent_upper.clone(), new_upper).await;
+                    match cas_result {
+                        Ok(Ok(Ok(()))) => break 'retry_loop,
+                        Ok(Ok(Err(Upper(upper)))) => {
+                            recent_upper = upper;
+                        }
+                        Ok(Err(e)) => {
+                            panic!("Invalid usage of the persist client for source {source_id} status history shard: {e:?}");
+                        }
+                        Err(e) => {
+                            trace!("compare_and_append in update_status failed: {e}");
```

Comment:
This is the case where we could become more fancy if we wanted. We would have to sync up on what's currently in the status shard, see if our desired message is in there, and potentially retry. Future work, though, it seems. 👌

Reply:
Right. I'm pretty hesitant to pull down the entire contents of the collection, since it'll include all the writes from all the storaged processes, placing quadradic read volume on the shard.

This does introduce some risk of duplicates, though! An alternative would be to plumb all the results back to environmentd and maintain the state there, which would put a lot less weight on storaged. We can explore that if duplicate entries end up being a problem.

Reply:
👌